packages/core/src/StarknetWindowObject.ts

```diff
@@ -0,0 +1,123 @@
+
+export type WalletEventHandlers =
+  | AccountChangeEventHandler
+  | NetworkChangeEventHandler
+
+export type WalletEvents =
+  | {
+      type: "accountsChanged"
+      handler: AccountChangeEventHandler
+    }
+  | {
+      type: "networkChanged"
+      handler: NetworkChangeEventHandler
+    }
+
+// EIP-747:
+// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-747.md
+export interface WatchAssetParameters {
+  type: "ERC20" // The asset's interface, e.g. 'ERC20'
+  options: {
+    address: string // The hexadecimal StarkNet address of the token contract
+    symbol?: string // A ticker symbol or shorthand, up to 5 alphanumerical characters
+    decimals?: number // The number of asset decimals
+    image?: string // A string url of the token logo
+    name?: string // The name of the token - not in spec
+  }
+}
+
+// EIP-3085
+// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-3085.md
+
+export interface AddStarknetChainParameters {
+  id: string
+  chainId: string // A 0x-prefixed hexadecimal string
+  chainName: string
+  baseUrl: string
+  rpcUrl?: string
+  blockExplorerUrl?: string
+  accountImplementation?: string
+
+  nativeCurrency?: {
+    name: string
+    symbol: string // 2-6 characters long
+    decimals: 18
+  } // Currently ignored.
+  iconUrls?: string[] // Currently ignored.
+}
+
+export interface SwitchStarknetChainParameter {
+  chainId: string // A 0x-prefixed hexadecimal string
+}
+
+export type RpcMessage =
+  | {
+      type: "wallet_watchAsset"
+      params: WatchAssetParameters
+      result: boolean
+    }
+  | {
+      type: "wallet_addStarknetChain"
+      params: AddStarknetChainParameters
+      result: boolean
+    }
+  | {
+      type: "wallet_switchStarknetChain"
+      params: SwitchStarknetChainParameter
+      result: boolean
+    }
+  | {
+      type: string
+      params: any
+      result: never
+    }
+
+interface IStarknetWindowObject {
+  id: string
+  name: string
+  version: string
+  icon: string
+  request: <T extends RpcMessage>(
+    call: Omit<T, "result">,
+  ) => Promise<T["result"]>
+  enable: (options?: { starknetVersion?: "v3" | "v4" }) => Promise<string[]>
+  isPreauthorized: () => Promise<boolean>
+  on: (
+    event: WalletEvents["type"],
+    handleEvent: WalletEvents["handler"],
+  ) => void
+  off: (
+    event: WalletEvents["type"],
+    handleEvent: WalletEvents["handler"],
+  ) => void
+  account?: AccountInterface
+  provider: ProviderInterface
+  selectedAddress?: string
+  chainId?: string
+}
+
+export interface ConnectedStarknetWindowObject extends IStarknetWindowObject {
+  isConnected: true
+  account: AccountInterface
+  selectedAddress: string
+  chainId: string
+}
+
+interface DisconnectedStarknetWindowObject extends IStarknetWindowObject {
+  isConnected: false
+}
+
+export type StarknetWindowObject =
+  | ConnectedStarknetWindowObject
+  | DisconnectedStarknetWindowObject
+
+declare global {
+  interface Window {
+    starknet?: StarknetWindowObject
```

Comment:
we should either remove this one, or we should add -
```
starknet_braavos?: StarknetWindowObject
```
in here (my suggestion - let's remove it)

Reply:
Added `[key: `starknet_${string}`]: StarknetWindowObject | undefined` good catch

I think we should keep it, as it allows dapps to simply access the properties in typescript without overwriting their global Window object

Reply:
ok, I understand. good suggestion with the generic typescript key.
still if we have an explicit key for one wallet it would be better to add other known keys to be used by IDE/auto-complete engines